src/Core/Cart/CartRuleCalculator.php

```diff
@@ -122,65 +122,67 @@ protected function applyCartRule(CartRuleData $cartRuleData, $withFreeShipping =
         }
 
         // Discount (%) on the whole order
-        if ($cartRule->reduction_percent && $cartRule->reduction_product == 0) {
-            foreach ($this->cartRows as $cartRow) {
-                $product = $cartRow->getRowData();
-                if ((($cartRule->reduction_exclude_special && !$product['reduction_applies'])
+        if ((float) $cartRule->reduction_percent > 0) {
+            if ($cartRule->reduction_product == 0) {
+                foreach ($this->cartRows as $cartRow) {
+                    $product = $cartRow->getRowData();
+                    if ((($cartRule->reduction_exclude_special && !$product['reduction_applies'])
                         || !$cartRule->reduction_exclude_special)) {
-                    $amount = $cartRow->applyPercentageDiscount($cartRule->reduction_percent);
-                    $cartRuleData->addDiscountApplied($amount);
+                        $amount = $cartRow->applyPercentageDiscount($cartRule->reduction_percent);
+                        $cartRuleData->addDiscountApplied($amount);
+                    }
                 }
             }
-        }
 
-        // Discount (%) on a specific product
-        if ($cartRule->reduction_percent && $cartRule->reduction_product > 0) {
-            foreach ($this->cartRows as $cartRow) {
-                if ($cartRow->getRowData()['id_product'] == $cartRule->reduction_product) {
-                    $amount = $cartRow->applyPercentageDiscount($cartRule->reduction_percent);
-                    $cartRuleData->addDiscountApplied($amount);
+            // Discount (%) on a specific product
+            if ($cartRule->reduction_product > 0) {
+                foreach ($this->cartRows as $cartRow) {
+                    if ($cartRow->getRowData()['id_product'] == $cartRule->reduction_product) {
+                        $amount = $cartRow->applyPercentageDiscount($cartRule->reduction_percent);
+                        $cartRuleData->addDiscountApplied($amount);
+                    }
                 }
             }
-        }
 
-        // Discount (%) on the cheapest product
-        if ($cartRule->reduction_percent && $cartRule->reduction_product == -1) {
-            /** @var CartRow|null $cartRowCheapest */
-            $cartRowCheapest = null;
-            foreach ($this->cartRows as $cartRow) {
-                $product = $cartRow->getRowData();
-                if (((($cartRule->reduction_exclude_special && !$product['reduction_applies'])
-                        || !$cartRule->reduction_exclude_special)) && ($cartRowCheapest === null
-                        || $cartRowCheapest->getInitialUnitPrice()->getTaxIncluded() > $cartRow->getInitialUnitPrice()
-                            ->getTaxIncluded())
-                ) {
-                    $cartRowCheapest = $cartRow;
+            // Discount (%) on the cheapest product
+            if ($cartRule->reduction_product == -1) {
```

Comment:
Can we improve a little bit the code and replace this `-1` by something that has meaning, like either a variable or a constant ?

Reply:
this part was just refactored, I'm not sure to find quickly what these values are meant to be